sdk/keyvault/azure_security_keyvault_secrets/src/resource.rs

```diff
@@ -0,0 +1,192 @@
+// Licensed under the MIT License.
+
+#[cfg(doc)]
+use crate::SecretClient;
+use azure_core::{error::ErrorKind, Result, Url};
+
+/// Information about the resource.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct ResourceId {
+    /// The complete [`Url`] of the resource.
+    pub id: Url,
+
+    /// The name of the resource.
+    pub name: String,
+
+    /// The optional version of the resource.
+    pub version: Option<String>,
+}
+
+/// Extension methods to get a [`ResourceId`] from models in this crate.
+pub trait ResourceExt {
+    /// Gets the [`ResourceId`] from this model.
+    ///
+    /// You can parse the name and version to pass to subsequent [`SecretClient`] method calls.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use azure_security_keyvault_secrets::{models::SecretBundle, ResourceExt as _};
+    ///
+    /// # fn main() -> Result<(), Box<dyn std::error::Error>> {
+    /// // SecretClient::get_secret() will return a SecretBundle.
+    /// let mut secret = SecretBundle::default();
+    /// secret.id = Some("https://my-vault.vault.azure.net/secrets/my-secret/abcd1234?api-version=7.5".into());
+    ///
+    /// let id = secret.resource_id()?;
+    /// assert_eq!(id.name, "my-secret");
+    /// assert_eq!(id.version, Some("abcd1234".into()));
+    /// # Ok(())
+    /// # }
+    /// ```
+    fn resource_id(&self) -> Result<ResourceId>;
+}
+
+impl<T> ResourceExt for T
+where
+    T: private::AsId,
+{
+    fn resource_id(&self) -> Result<ResourceId> {
+        let Some(id) = self.as_id() else {
+            return Err(azure_core::Error::message(
+                ErrorKind::DataConversion,
+                "missing resource id",
+            ));
+        };
+
+        let url: Url = id.parse()?;
+        deconstruct(url)
+    }
+}
+
+fn deconstruct(url: Url) -> Result<ResourceId> {
+    let mut segments = url
+        .path_segments()
+        .ok_or_else(|| azure_core::Error::message(ErrorKind::DataConversion, "invalid url"))?;
+    segments
+        .next()
+        .and_then(none_if_empty)
+        .ok_or_else(|| azure_core::Error::message(ErrorKind::DataConversion, "missing collection"))
+        .and_then(|col| {
+            if col != "secrets" {
+                return Err(azure_core::Error::message(
+                    ErrorKind::DataConversion,
+                    "not in secrets collection",
+                ));
+            }
+            Ok(col)
+        })?;
+    let name = segments
+        .next()
+        .and_then(none_if_empty)
+        .ok_or_else(|| azure_core::Error::message(ErrorKind::DataConversion, "missing name"))
+        .map(String::from)?;
+    let version = segments.next().and_then(none_if_empty).map(String::from);
+
+    Ok(ResourceId {
+        id: url,
+        name,
+        version,
+    })
+}
+
+fn none_if_empty(s: &str) -> Option<&str> {
+    if s.is_empty() {
+        return None;
+    }
+
+    Some(s)
+}
+
+mod private {
```

Comment:
This is a common practice to prevent others from implementing the trait - making either a base trait or required trait private so only our module can implement it.